tfhe/src/core_crypto/commons/math/torus/mod.rs

```diff
@@ -83,26 +78,45 @@ macro_rules! implement {
             }
             #[inline]
-            // WARNING from documentation reproduced here
-            // It is the caller's reponsibility to provide a custom_modulus that is a safe
-            // approximation of the integer modulus they want to use in the floating point
-            // domain.
-            //
-            // If the approximate floating point modulus is too big then some values will be out of
-            // the proper range for the given integer modulus.
-            fn from_torus_custom_modulus(input: F, custom_modulus: F) -> Self {
+            fn from_torus_custom_mod(input: F, custom_modulus: Self) -> Self {
+                // Reasonging with F = f64
+                //
+                // Here we convert the custom modulus to f64 but rounding down to the closest
+                // representable value by f64. This allows to make sure that the value that
+                // is outputed is in the correct range with respect to our modulus.
+                // The max error for a u64 is 2048, i.e. max(diff(floor(u64 as f64) as u64, u64)) <
+                // 2048 This small error should be acceptable for values produced by the
+                // gaussian generation as they have to be related to the modulus itself,
+                // 2048 / 2^63 = 2^-52, basically the f64 relative error is the max relative
+                // error you can get. This is valid for other moduli >= 2^53 with a max
+                // absolute error as low as 1 e.g. for 2^64 - 2^32 + 1, which rounds down to
+                // 2^64 - 2^32 in the float domain.
+                let custom_modulus_float: F =
+                    convert_unsigned_integer_to_float_truncate(custom_modulus);
+
                 // This is in [-0.5, 0.5[
+                // We do not do the mapping to [0, 1[ here as some values can be extremely small
+                // (think 2^-127 for u128 and custom power of 2 moduli) and would be crushed by
+                // adding 1 to them, creating artificial zeros (not good for noise generation)
                 let mut fract = input - F::round(input);
-                // Scale to the modulus
-                fract *= custom_modulus;
-                // This allows to map the negative part of the [-0.5, 0.5[ torus to the upper part
-                // of the [0, 1[ torus in a single operation.
-                // Also this is better done here to avoid epsilon issues as input can be very very
-                // small (i.e. if input is very small adding 1.0 to input to map it to the positive
-                // values would always produce 1.0, and therefore a noise of 0.0 if this value is
-                // used for noise generation, big yikes)
-                fract = F::round(fract).rem_euclid(custom_modulus);
-                return fract.cast_into();
+                // Scale to the truncated modulus
+                fract *= custom_modulus_float;
+                fract = F::round(fract);
```

Comment:
May be safer to do a round_towards_infinity

Reply:
https://github.com/zama-ai/tfhe-rs-internal/issues/326